File: battery_analysis_v2/core/dvdq_analysis/fullcell_fitting.py
"""
Full-Cell Fitting Module
========================

Full-cell 전압 시뮬레이션 및 dV/dQ 계산 모듈입니다.

원본 함수: BatteryDataTool.py - generate_simulation_full()

핵심 수식:
    V_fullcell(Q) = V_cathode(Q + slip_ca) * mass_ca - V_anode(Q + slip_an) * mass_an
    dV/dQ = d(V_fullcell)/dQ

Author: Battery Analysis Team
Date: 2026-01-14
"""

import numpy as np
import pandas as pd
from typing import Tuple, Optional, Union
from dataclasses import dataclass



from .halfcell_profiles import HalfCellProfile


@dataclass
class FittingResult:
    """피팅 결과 데이터 구조"""
    ca_mass: float
    ca_slip: float
    an_mass: float
    an_slip: float
    rmse: float
    r_squared: float
    simulated_voltage: np.ndarray
    simulated_dvdq: np.ndarray


def calculate_dvdq(voltage: np.ndarray, 
                   capacity: np.ndarray, 
                   period: int = 1) -> np.ndarray:
    """
    dV/dQ 계산 (미분 전압)
    
    Args:
        voltage: 전압 데이터 (V)
        capacity: 용량 데이터 (mAh 또는 %)
        period: 미분 윈도우 크기
        
    Returns:
        dV/dQ 배열
    """
    dv = np.diff(voltage, n=period, prepend=np.nan * np.ones(period))
    dq = np.diff(capacity, n=period, prepend=np.nan * np.ones(period))
    
    # 0으로 나누기 방지
    with np.errstate(divide='ignore', invalid='ignore'):
        dvdq = dv / dq
        dvdq[~np.isfinite(dvdq)] = np.nan
    
    return dvdq


def generate_simulation_full(
    ca_capacity: np.ndarray,
    ca_voltage: np.ndarray,
    an_capacity: np.ndarray,
    an_voltage: np.ndarray,
    real_capacity: np.ndarray,
    real_voltage: np.ndarray,
    ca_mass: float,
    ca_slip: float,
    an_mass: float,
    an_slip: float,
    full_cell_max_cap: float,
    rated_cap: float,
    full_period: int = 1
) -> pd.DataFrame:
    """
    Full-cell 시뮬레이션 결과 생성
    
    원본: BatteryDataTool.py generate_simulation_full() 함수
    
    열화 모델:
        ca_cap_new = ca_cap * ca_mass - ca_slip
        an_cap_new = an_cap * an_mass - an_slip
        V_full = V_cathode(Q) - V_anode(Q)
    
    Args:
        ca_capacity: 양극 용량 데이터
        ca_voltage: 양극 전압 데이터
        an_capacity: 음극 용량 데이터
        an_voltage: 음극 전압 데이터
        real_capacity: 실측 용량 데이터
        real_voltage: 실측 전압 데이터
        ca_mass: 양극 활물질 비율 (0-1)
        ca_slip: 양극 SOC 슬립
        an_mass: 음극 활물질 비율 (0-1)
        an_slip: 음극 SOC 슬립
        full_cell_max_cap: 풀셀 최대 용량
        rated_cap: 정격 용량
        full_period: dV/dQ 계산 주기
        
    Returns:
        시뮬레이션 결과 DataFrame
    """
    # 용량 보정 적용
    ca_cap_new = ca_capacity * ca_mass - ca_slip
    an_cap_new = an_capacity * an_mass - an_slip
    
    # 기준 용량 생성 (0.1 mAh 단위)
    simul_full_cap = np.arange(0, full_cell_max_cap, 0.1)
    
    # 각 전극 전압 보간
    simul_full_ca_volt = np.interp(simul_full_cap, ca_cap_new, ca_voltage)
    simul_full_an_volt = np.interp(simul_full_cap, an_cap_new, an_voltage)
    simul_full_real_volt = np.interp(simul_full_cap, real_capacity, real_voltage)
    
    # Full-cell 전압 계산
    simul_full_volt = simul_full_ca_volt - simul_full_an_volt
    
    # DataFrame 생성
    simul_full = pd.DataFrame({
        "full_cap": simul_full_cap,
        "an_volt": simul_full_an_volt,
        "ca_volt": simul_full_ca_volt,
        "full_volt": simul_full_volt,
        "real_volt": simul_full_real_volt
    })
    
    # 마지막 행 제거 (경계 조건)
    simul_full = simul_full.iloc[:-1]
    
    # 백분율로 용량 변환
    simul_full["full_cap"] = simul_full["full_cap"] / rated_cap * 100
    
    # dV/dQ 계산
    simul_full["an_dvdq"] = calculate_dvdq(
        simul_full["an_volt"].values, 
        simul_full["full_cap"].values, 
        full_period
    )
    simul_full["ca_dvdq"] = calculate_dvdq(
        simul_full["ca_volt"].values,
        simul_full["full_cap"].values,
        full_period
    )
    simul_full["real_dvdq"] = calculate_dvdq(
        simul_full["real_volt"].values,
        simul_full["full_cap"].values,
        full_period
    )
    simul_full["full_dvdq"] = simul_full["ca_dvdq"] - simul_full["an_dvdq"]
    
    return simul_full


class FullCellSimulator:
    """
    Full-cell 시뮬레이션 클래스
    
    사용 예시:
        >>> cathode = HalfCellProfile(...)
        >>> anode = HalfCellProfile(...)
        >>> simulator = FullCellSimulator(cathode, anode)
        >>> result = simulator.simulate(ca_mass=0.95, an_mass=0.98, ...)
    """
    
    def __init__(self, 
                 cathode_profile: HalfCellProfile,
                 anode_profile: HalfCellProfile,
                 rated_capacity: float = 100.0):
        """
        Args:
            cathode_profile: 양극 OCV 프로파일
            anode_profile: 음극 OCV 프로파일
            rated_capacity: 정격 용량 (mAh)
        """
        self.cathode = cathode_profile
        self.anode = anode_profile
        self.rated_capacity = rated_capacity
    
    def simulate(self,
                 ca_mass: float = 1.0,
                 ca_slip: float = 0.0,
                 an_mass: float = 1.0,
                 an_slip: float = 0.0,
                 capacity_step: float = 0.1,
                 dvdq_period: int = 1) -> pd.DataFrame:
        """
        Full-cell 전압 시뮬레이션 수행
        
        Args:
            ca_mass: 양극 활물질 비율
            ca_slip: 양극 SOC 슬립
            an_mass: 음극 활물질 비율
            an_slip: 음극 SOC 슬립
            capacity_step: 용량 스텝 크기
            dvdq_period: dV/dQ 미분 주기
            
        Returns:
            시뮬레이션 결과 DataFrame
        """
        # 열화 파라미터 적용
        ca_cap_new, ca_volt = self.cathode.apply_degradation(ca_mass, ca_slip)
        an_cap_new, an_volt = self.anode.apply_degradation(an_mass, an_slip)
        
        # 최대 용량 결정
        max_cap = min(np.max(ca_cap_new), np.max(an_cap_new))
        
        # 시뮬레이션 용량 범위
        sim_capacity = np.arange(0, max_cap, capacity_step)
        
        # 전압 보간
        sim_ca_volt = np.interp(sim_capacity, ca_cap_new, ca_volt)
        sim_an_volt = np.interp(sim_capacity, an_cap_new, an_volt)
        
        # Full-cell 전압
        sim_full_volt = sim_ca_volt - sim_an_volt
        
        # SOC 계산 (백분율)
        sim_soc = sim_capacity / self.rated_capacity * 100
        
        # DataFrame 생성
        result = pd.DataFrame({
            'capacity': sim_capacity,
            'soc': sim_soc,
            'ca_voltage': sim_ca_volt,
            'an_voltage': sim_an_volt,
            'full_voltage': sim_full_volt
        })
        
        # dV/dQ 계산
        result['ca_dvdq'] = calculate_dvdq(sim_ca_volt, sim_soc, dvdq_period)
        result['an_dvdq'] = calculate_dvdq(sim_an_volt, sim_soc, dvdq_period)
        result['full_dvdq'] = calculate_dvdq(sim_full_volt, sim_soc, dvdq_period)
        
        return result
    
    def calculate_rmse(self,
                       experimental_voltage: np.ndarray,
                       experimental_capacity: np.ndarray,
                       ca_mass: float,
                       ca_slip: float,
                       an_mass: float,
                       an_slip: float) -> float:
        """
        시뮬레이션과 실험 데이터 간 RMSE 계산
        
        Args:
            experimental_voltage: 실험 전압 데이터
            experimental_capacity: 실험 용량 데이터
            ca_mass, ca_slip, an_mass, an_slip: 열화 파라미터
            
        Returns:
            RMSE 값
        """
        # 시뮬레이션 수행
        sim_result = self.simulate(ca_mass, ca_slip, an_mass, an_slip)
        
        # 실험 용량 범위에서 시뮬레이션 전압 보간
        sim_voltage_interp = np.interp(
            experimental_capacity,
            sim_result['capacity'].values,
            sim_result['full_voltage'].values
        )
        
        # RMSE 계산
        rmse = np.sqrt(np.mean((experimental_voltage - sim_voltage_interp) ** 2))
        
        return rmse

File: julia/src/DVDQAnalysis/FullCellFitting.jl
"""
FullCellFitting.jl - Full-cell dV/dQ 피팅
==========================================

Python fullcell_fitting.py의 Julia 버전입니다.

핵심 수식:
    V_fullcell(Q) = V_cathode(Q) - V_anode(Q)
    dV/dQ = d(V_fullcell)/dQ
"""

"""
    calculate_dvdq(voltage, capacity; period=1)

dV/dQ 계산 (미분 전압)

# Arguments
- `voltage`: 전압 배열 (V)
- `capacity`: 용량 배열 (mAh 또는 %)
- `period`: 미분 윈도우 크기

# Returns
- dV/dQ 배열
"""
function calculate_dvdq(voltage::Vector{T}, capacity::Vector{T};
    period::Int=1)::Vector{T} where T<:AbstractFloat
    n = length(voltage)
    result = Vector{T}(undef, n)

    @inbounds for i in 1:n
        if i <= period
            result[i] = NaN
        else
            dv = voltage[i] - voltage[i-period]
            dq = capacity[i] - capacity[i-period]
            result[i] = dq != 0 ? dv / dq : NaN
        end
    end

    return result
end


"""
    simulate_fullcell(ca_cap, ca_volt, an_cap, an_volt, params::FittingParams;
                      max_cap=100.0, step=0.1)

Full-cell 전압 시뮬레이션

# Arguments
- `ca_cap, ca_volt`: 양극 용량/전압 데이터
- `an_cap, an_volt`: 음극 용량/전압 데이터
- `params`: 열화 파라미터 (FittingParams)
- `max_cap`: 최대 용량
- `step`: 용량 스텝

# Returns
- (capacity, ca_voltage, an_voltage, full_voltage, full_dvdq) 튜플
"""
function simulate_fullcell(ca_cap::Vector{T}, ca_volt::Vector{T},
    an_cap::Vector{T}, an_volt::Vector{T},
    params::FittingParams;
    max_cap::T=100.0, step::T=0.1) where T<:AbstractFloat

    # 열화 적용
    ca_cap_new = ca_cap .* params.ca_mass .- params.ca_slip
    an_cap_new = an_cap .* params.an_mass .- params.an_slip

    # 시뮬레이션 용량 범위
    sim_capacity = collect(0.0:step:max_cap)
    n = length(sim_capacity)

    sim_ca_volt = Vector{T}(undef, n)
    sim_an_volt = Vector{T}(undef, n)
    sim_full_volt = Vector{T}(undef, n)

    # 보간 수행
    @inbounds for i in 1:n
        q = sim_capacity[i]
        sim_ca_volt[i] = linear_interp(q, ca_cap_new, ca_volt)
        sim_an_volt[i] = linear_interp(q, an_cap_new, an_volt)
        sim_full_volt[i] = sim_ca_volt[i] - sim_an_volt[i]
    end

    # dV/dQ 계산
    sim_full_dvdq = calculate_dvdq(sim_full_volt, sim_capacity)

    return (sim_capacity, sim_ca_volt, sim_an_volt, sim_full_volt, sim_full_dvdq)
end


"""
    linear_interp(x, xp, fp)

선형 보간 (수동 구현)
"""
@inline function linear_interp(x::T, xp::Vector{T}, fp::Vector{T})::T where T<:AbstractFloat
    n = length(xp)

    # 경계 처리
    if x <= xp[1]
        return fp[1]
    elseif x >= xp[n]
        return fp[n]
    end

    # 이진 검색
    lo, hi = 1, n
    while lo < hi - 1
        mid = (lo + hi) ÷ 2
        if xp[mid] <= x
            lo = mid
        else
            hi = mid
        end
    end

    # 선형 보간
    t = (x - xp[lo]) / (xp[hi] - xp[lo])
    return fp[lo] + t * (fp[hi] - fp[lo])
end


"""
    calculate_rmse(exp_voltage, exp_capacity, ca_cap, ca_volt, an_cap, an_volt, params)

시뮬레이션과 실험 데이터 간 RMSE 계산
"""
function calculate_rmse(exp_voltage::Vector{T}, exp_capacity::Vector{T},
    ca_cap::Vector{T}, ca_volt::Vector{T},
    an_cap::Vector{T}, an_volt::Vector{T},
    params::FittingParams) where T<:AbstractFloat

    # 시뮬레이션 수행
    sim_cap, sim_ca, sim_an, sim_full, _ = simulate_fullcell(
        ca_cap, ca_volt, an_cap, an_volt, params;
        max_cap=maximum(exp_capacity)
    )

    # 실험 용량에서 시뮬레이션 전압 보간
    sim_interp = [linear_interp(q, sim_cap, sim_full) for q in exp_capacity]

    # RMSE 계산
    rmse = sqrt(mean((exp_voltage .- sim_interp) .^ 2))

    return rmse
end
